src/plugins/spatial_grid.rs

```diff
@@ -0,0 +1,482 @@
+            })
+    }
+
+    /// Clear all entities from the spatial index.
+    pub fn clear(&mut self)
+    {
+        self.position_to_entities.clear();
+        self.entity_to_position.clear();
+    }
+
+    /// Check if a position is empty (has no entities).
+    #[must_use]
+    pub fn is_empty(&self, position: &GridPosition<T>) -> bool
+    {
+        self.position_to_entities
+            .get(position)
+            .is_none_or(HashSet::is_empty)
+    }
+
+    /// Get total number of entities in the grid.
+    #[must_use]
+    pub fn num_entities(&self) -> usize
+    {
+        self.entity_to_position.len()
+    }
+}
+
+/// Plugin that maintains a spatial index for entities with `GridPosition` components.
+/// Generic over coordinate types that implement the `GridCoordinate` trait and component types.
+pub struct SpatialGridPlugin<T: GridCoordinates, C: Component>
+{
+    bounds: Option<GridBounds<T>>,
+    _phantom: std::marker::PhantomData<(T, C)>,
+}
+
+impl<T: GridCoordinates, C: Component> SpatialGridPlugin<T, C>
+{
+    pub const fn new(bounds: Option<GridBounds<T>>) -> Self
+    {
+        Self {
+            bounds,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+
+    pub fn init(app: &mut App, bounds: Option<GridBounds<T>>)
+    {
+        // Spawn the spatial grid entity directly
+        let spatial_grid = SpatialGrid::<T, C>::new(bounds);
+        app.world_mut().insert_resource(spatial_grid);
+    }
+}
+
+impl<T: GridCoordinates, C: Component> Plugin for SpatialGridPlugin<T, C>
+{
+    fn build(&self, app: &mut App)
+    {
+        Self::init(app, self.bounds);
+
+        // System to maintain the spatial index
+        app.add_systems(
+            PreUpdate,
+            (
+                spatial_grid_reset_system::<T, C>,
+                spatial_grid_update_system::<T, C>,
+                spatial_grid_cleanup_system::<T, C>,
+            )
+                .chain(),
+        );
+    }
+}
+
+/// System that resets the spatial grid at the beginning of each simulation.
+fn spatial_grid_reset_system<T: GridCoordinates, C: Component>(
+    mut spatial_grid: ResMut<SpatialGrid<T, C>>,
+    step_counter: Res<StepCounter>,
+)
+{
+    // Reset the spatial grid whenever the step counter is 0
+    // This should occur on the first step of every simulation
+    if **step_counter == 0
+    {
+        spatial_grid.clear();
+    }
+}
+
+/// Query for entities with `GridPosition` components that have been added or changed.
+type GridPositionQuery<'world, 'state, T, C> =
+    Query<'world, 'state, (Entity, &'static GridPosition<T>), (Changed<GridPosition<T>>, With<C>)>;
+
+/// System that updates the spatial grid when entities with `GridPosition` are added or moved.
+fn spatial_grid_update_system<T: GridCoordinates, C: Component>(
+    mut spatial_grid: ResMut<SpatialGrid<T, C>>,
+    query: GridPositionQuery<T, C>,
+)
+{
+    for (entity, position) in &query
+    {
+        spatial_grid.remove(entity);
```

Comment:
Sorry my bad, I didn't see that `insert()` calls `remove()` on its own.
Can remove this line.